src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+
+        const auto plateaux = parent_parent_in_forward == via && parent_parent_in_reverse == via;
+
+        return !plateaux;
+    };
+
+    return std::remove_if(first, last, is_not_locally_optimal);
+}
+
+// The unpacked paths' sharing in [0, 1] for no sharing and equality, respectively.
+inline double normalizedUnpackedPathSharing(const UnpackedNodes &lhs, const UnpackedNodes &rhs)
+{
+    if (lhs.empty() || rhs.empty())
+        return 0.;
+
+    // Todo: stack alloc?
+    std::vector<NodeID> lhs_nodes{begin(lhs), end(lhs)};
+    std::vector<NodeID> rhs_nodes{begin(rhs), end(rhs)};
+
+    std::sort(begin(lhs_nodes), end(lhs_nodes));
+    std::sort(begin(rhs_nodes), end(rhs_nodes));
+
+    std::size_t num_different = 0;
+
+    // Needs to be wrapped in std::function for Windows compiler bug
+    auto out = boost::make_function_output_iterator(
+        std::function<void(NodeID)>([&](auto) { num_different += 1; }));
+
+    std::set_difference(begin(lhs_nodes), end(lhs_nodes), begin(rhs_nodes), end(rhs_nodes), out);
+
+    const auto difference = static_cast<double>(num_different) / lhs.size();
+
+    BOOST_ASSERT(difference >= 0.);
+    BOOST_ASSERT(difference <= 1.);
+
+    // Todo: do we need to scale sharing with edge weights in some sort?
+
+    return 1. - difference;
+}
+
+// Filters unpacked paths compared to the primary route. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt
+filterUnpackedPathsBySharing(const WeightedViaNodeUnpackedPath &path, RandIt first, RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNodeUnpackedPath>();
+
+    const auto over_sharing_limit = [&](const auto &unpacked) {
+        return normalizedUnpackedPathSharing(path.nodes, unpacked.nodes) > kGamma;
+    };
+
+    return std::remove_if(first, last, over_sharing_limit);
+}
+
+} // anon. ns
+
+// Alternative Routes for MLD.
+//
+// Start search from s and continue "for a while" when midd was found. Save all vertices.
+// Start search from t and continue "for a while" when s was found. Save all vertices.
+// Intersect both vertex sets: these are the candidate vertices.
+// For all candidate vertices c a (potentially arbitrarily bad) alternative route is (s, c, t).
+// Apply heuristic to evaluate alternative route based on stretch, overlap, how reasonable it is.
+//
+// For MLD specifically we can pull off some tricks to make evaluating alternatives fast:
+//   Only consider (s, c, t) with c border vertex: re-use MLD search steps.
+//   Add meta data to border vertices: consider (s, c, t) only when c is e.g. on a highway.
+//   Prune based on vertex cell id
+//
+// https://github.com/Project-OSRM/osrm-backend/issues/3905
+InternalManyRoutesResult
+alternativePathSearch(SearchEngineData<Algorithm> &search_engine_data,
```

Comment:
This is becoming a 👹  of a function. Do you think it could be split into multiple sub-functions?